sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -39,81 +39,161 @@
 
 class SdkHarness(object):
 
-  def __init__(self, control_address):
+  def __init__(self, control_address, pipeline_options=None):
+
+    def _get_worker_count(pipeline_options):
+      experiments = pipeline_options.get(
+          'experiments'
+      ) if pipeline_options and pipeline_options.has_key('experiments') else []
+
+      experiments = experiments if experiments else []
+
+      for experiment in experiments:
+        # There should only be 1 match so returning from the loop
+        s = re.findall('^(worker_threads=[0-9]+)', experiment)
+        if len(s) > 0:
+          return int(s.pop()[len('worker_threads='):])
+
+      return 1
+
+    self._worker_count = _get_worker_count(pipeline_options)
+    logging.info('Initializing SDKHarness with %s workers.', self._worker_count)
+    self._worker_index = 0
     self._control_channel = grpc.insecure_channel(control_address)
     self._data_channel_factory = data_plane.GrpcClientDataChannelFactory()
-    # TODO: Ensure thread safety to run with more than 1 thread.
-    self._default_work_thread_pool = futures.ThreadPoolExecutor(max_workers=1)
+    self.worker_wrappers = []
+    # one thread is enough for getting the progress report
```

Comment:
<!--new_thread; commit:5dfb3c0f3b174719be925a98b29ef1648d74545c; resolved:0-->
Is this true for any number of worker threads?